lib/util.js

```diff
@@ -1057,3 +1057,72 @@ exports._exceptionWithHostPort = function(err,
 // process.versions needs a custom function as some values are lazy-evaluated.
 process.versions[exports.inspect.custom] =
   (depth) => exports.format(JSON.parse(JSON.stringify(process.versions)));
+
+
+/**
+ * @typedef {Object} callbackifyOptions
+ * @property {function} transformer
+ * @property {boolean} toSplat
+ * @property {Array.<String>} project
+ */
+
+/**
+ * @param {function(): Promise} orig
+ * @param {callbackifyOptions} options
+ */
+function callbackify(orig, options) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to callbackify() ' +
+                        'must be a function');
+  }
+
+  // we have two way to convert the returned value into multiple arguments
+  // either splat it or project some of it's keys
+  const toSplat = options.toSplat;
+  const project = options.project;
+
+  const fn = function(...args) {
+    const cb = args.pop();
+    if (typeof cb !== 'function') {
+      throw new TypeError('The first argument to the callbackified function ' +
+        'must be a callback style function');
+    }
+    var p = orig.call(this, ...args);
+    if (typeof options.transformer === 'function') {
+      p = options.transformer(p);
+    }
+    p = p.then((ret) => {
+      if (toSplat) {
+        cb(null, ...ret);
+      } else if (project) {
+        const allThere = project.all((k) => k in ret);
+        if (!allThere) {
+          const notAll = new TypeError('Not all of the projected properties' +
+                                       ' are present in the returned value:' +
+                                       inspect(ret));
+          return cb(notAll);
+        } else {
+          const args = project.reduce((s, k) => s.concat(ret[k]), []);
+          cb(null, ...args);
+        }
+      } else {
+        cb(null, ret);
+      }
+    }, (rej) => {
+      // !rej guard inspired by BlueBird https://goo.gl/t5IS6M
```

Comment:
To clarify, I didn't suggest doing this so that errors would have a stack trace -- I suggested it because `null` is a special error value in callbacks which means "no error occurred". If we don't error-wrap, there is no way for a callback consumer to distinguish between "the promise rejected with `null`" and "the promise fulfilled with `undefined`".